test/Validate.CFG.cpp

```diff
@@ -0,0 +1,629 @@
+  Block block3("block3");
+  Block def("def");  // default block
+  Block merge("merge");
+  Block end("end", SpvOpReturn);
+
+  bad.setBody(
+      "%cond    = OpSLessThan %intt %one %two\n"
+      "OpSelectionMerge %merge None\n");
+
+  string str = header + nameOps("entry", "bad", make_pair("func", "Main")) +
+               types_consts + "%func    = OpFunction %voidt None %funct\n";
+
+  str += entry >> bad;
+  str += bad >> vector<Block>({def, block1, block2, block3, entry});
+  str += def >> merge;
+  str += block1 >> merge;
+  str += block2 >> merge;
+  str += block3 >> merge;
+  str += merge >> end;
+  str += end;
+  str += "OpFunctionEnd\n";
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_ERROR_INVALID_CFG, ValidateInstructions());
+  EXPECT_THAT(
+      getDiagnosticString(),
+      MatchesRegex("First block [0-9]\\[entry\\] of funciton [0-9]\\[Main\\] "
+                   "is targeted by block [0-9]\\[bad\\]"));
+}
+
+TEST_F(ValidateCFG, BranchToBlockInOtherFunctionBad) {
+  Block entry("entry");
+  Block middle("middle", SpvOpBranchConditional);
+  Block end("end", SpvOpReturn);
+
+  middle.setBody(
+      "%cond    = OpSLessThan %intt %one %two\n"
+      "OpSelectionMerge %end None\n");
+
+  Block entry2("entry2");
+  Block middle2("middle2");
+  Block end2("end2", SpvOpReturn);
+
+  string str = header + nameOps("middle2", make_pair("func", "Main")) +
+               types_consts + "%func    = OpFunction %voidt None %funct\n";
+
+  str += entry >> middle;
+  str += middle >> vector<Block>({end, middle2});
+  str += end;
+  str += "OpFunctionEnd\n";
+
+  str += "%func2    = OpFunction %voidt None %funct\n";
+  str += entry2 >> middle2;
+  str += middle2 >> end2;
+  str += end2;
+  str += "OpFunctionEnd\n";
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_ERROR_INVALID_CFG, ValidateInstructions());
+  EXPECT_THAT(
+      getDiagnosticString(),
+      MatchesRegex(
+          "Block\\(s\\) \\{[0-9]\\[middle2\\] .\\} are referenced but not "
+          "defined in function [0-9]\\[Main\\]"));
+}
+
+TEST_F(ValidateCFG, HeaderDoesntDominatesMergeBad) {
+  Block entry("entry", SpvOpBranchConditional);
+  Block loop_head("loop_head");
+  Block branch("branch", SpvOpBranchConditional);
+  Block t("t");
+  Block f("f");
+  Block entry_merge("entry_merge");
+  Block loop_merge("loop_merge");
+  Block branch_merge("branch_merge");
+  Block exit("exit", SpvOpReturn);
+
+  entry.setBody(
+      "%cond = OpSLessThan %intt %one %two\n"
+      "OpSelectionMerge %entry_merge None\n");
+
+  branch.setBody("OpSelectionMerge %branch_merge None\n");
+
+  loop_head.setBody("OpLoopMerge %loop_merge %branch None\n");
+
+  string str = header + nameOps("loop_head", "loop_merge", make_pair("func", "Main")) + types_consts +
+               "%func    = OpFunction %voidt None %funct\n";
+
+  str += entry >> vector<Block>({loop_head, loop_merge});
+  str += loop_head >> branch;
+  str += branch >> vector<Block>({t, f});
+  str += t >> loop_head;
```

Comment:
Sigh, now the back-edge block is not post-dominating the continue target, violating another rule. :confounded: I'd go with something quite simple:
```
entry >> merge
header >> {merge, f}  // header is unreachable
```
I believe this violates only the rule about header dominating the merge.